[SPARK-4233] [SPARK-4367] [SPARK-3947] [SPARK-3056] [SQL] Aggregation Improvement

yhuai opened the pull request:
This is the first PR for the aggregation improvement, which is tracked by https://issues.apache.org/jira/browse/SPARK-4366 (umbrella JIRA). This PR contains work for its subtasks, SPARK-3056, SPARK-3947, SPARK-4233, and SPARK-4367.

This PR introduces a new code path for evaluating aggregate functions. This code path is guarded by `spark.sql.useAggregate2` and by default the value of this flag is true.

This new code path contains:
- A new aggregate function interface (`AggregateFunction2`) and 7 built-int aggregate functions based on this new interface (`AVG`, `COUNT`, `FIRST`, `LAST`, `MAX`, `MIN`, `SUM`)
- A UDAF interface (`UserDefinedAggregateFunction`) based on the new code path and two example UDAFs (`MyDoubleAvg` and `MyDoubleSum`).
- A sort-based aggregate operator (`Aggregate2Sort`) for the new aggregate function interface .
- A sort-based aggregate operator (`FinalAndCompleteAggregate2Sort`) for distinct aggregations (for distinct aggregations the query plan will use `Aggregate2Sort` and `FinalAndCompleteAggregate2Sort` together).

With this change, `spark.sql.useAggregate2` is `true`, the flow of compiling an aggregation query is:
1. Our analyzer looks up functions and returns aggregate functions built based on the old aggregate function interface.
2. When our planner is compiling the physical plan, it tries try to convert all aggregate functions to the ones built based on the new interface. The planner will fallback to the old code path if any of the following two conditions is true:
- code-gen is disabled.
- there is any function that cannot be converted (right now, Hive UDAFs).
- the schema of grouping expressions contain any complex data type.
- There are multiple distinct columns.

Right now, the new code path handles a single distinct column in the query (you can have multiple aggregate functions using that distinct column). For a query having a aggregate function with DISTINCT and regular aggregate functions, the generated plan will do partial aggregations for those regular aggregate function.

Thanks @chenghao-intel for his initial work on it.

yhuai committed:
wip

marmbrus committed:
Add AlgebraicAggregate

marmbrus committed:
now with correct answers\!

marmbrus committed:
First draft of codegen

yhuai committed:
Add an analysis rule to convert aggregate function to the new version.

yhuai committed:
Code cleanup: Remove unnecesary AttributeReferences.

marmbrus committed:
Merge remote-tracking branch 'apache/master' into UDAF

yhuai committed:
Bug fix: Handle expressions in grouping columns that are not attribute references.

yhuai committed:
Do not bind references in AlgebraicAggregate and use code gen for all places.

marmbrus committed:
add test

marmbrus committed:
make hive test

yhuai committed:
Add ConvertAggregateFunction to HiveContext's analyzer.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

yhuai committed:
Make Aggregate2Sort work with both algebraic AggregateFunctions and non-algebraic AggregateFunctions.

yhuai committed:
Bug fix.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

Conflicts:
	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

yhuai committed:
Cleanup.

yhuai committed:
Add CheckAggregateFunction to extendedCheckRules.

yhuai committed:
Remove groupOrdering.

yhuai committed:
Format update.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -146,6 +147,7 @@ object FunctionRegistry {
     expression[Max]("max"),
     expression[Min]("min"),
     expression[Sum]("sum"),
+    expression[MyDoubleSum]("mydoublesum"),
```
`MyDoubleSum` is just an example. It will be removed later.

SparkQA commented:
[Test build #37576 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37576/consoleFull) for   PR 7458 at commit [`6edb5ac`](https://github.com/apache/spark/commit/6edb5ace6a6939c8a46842d981a6b57ceb784db7).

SparkQA commented:
[Test build #37576 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37576/console) for   PR 7458 at commit [`6edb5ac`](https://github.com/apache/spark/commit/6edb5ace6a6939c8a46842d981a6b57ceb784db7).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class MyDoubleSum(child: Expression) extends AggregateFunction2 `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Aggregate2Sort(`
  * `case class ConvertAggregateFunction(context: SQLContext) extends Rule[LogicalPlan] `
  * `case class CheckAggregateFunction(context: SQLContext) extends (LogicalPlan => Unit) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

yhuai committed:
Make it compile.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -537,6 +541,7 @@ class Analyzer(
     def containsAggregates(exprs: Seq[Expression]): Boolean = {
       exprs.foreach(_.foreach {
         case agg: AggregateExpression => return true
```
if we are not going to be removing the old aggregate expression for a while, how about creating a new aggregateexpression that's the shared parent, and rename existing one aggregateexpression1? i worry we miss places that we match on.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -90,7 +90,7 @@ abstract class Expression extends TreeNode[Expression] {
     val primitive = ctx.freshName("primitive")
     val ve = GeneratedExpressionCode("", isNull, primitive)
     ve.code = genCode(ctx, ve)
-    ve
```
for very large expressions, this can be pretty bad. i wonder if we should have a debug mode.

marmbrus replied:
I think it is probably worth it.  When codegen bugs pop-up in the real world this makes it possible to debug, where as otherwise the code is super hard to follow.  I tested on a pretty deeply nested expression and the code size only went from 25kb to 50kb.  I'd imagine the parser is pretty good at skipping over comments.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala:
```diff
@@ -0,0 +1,365 @@
+  protected override def doExecute(): RDD[InternalRow] = attachTree(this, "execute") {
+    child.execute().mapPartitions { iter =>
+
+      new Iterator[InternalRow] {
```
we should take this iterator out as a named iterator

chenghao-intel replied:
Probably we'd better split the partial / final aggregation Iterator into different class. As the partial aggregation(pre-shuffle) can be followed by hash/sort aggregation(post-shuffle) for the further improvement.

yhuai replied:
@chenghao-intel Yeah, that's a good idea.

yhuai committed:
Add UDAF interface.

yhuai commented on sql/hive/src/test/java/test/org/apache/spark/sql/hive/aggregate2/MyJavaUDAF.java:
```diff
@@ -0,0 +1,100 @@
+      return buffer.getDouble(0);
+    }
+  }
+}
```
@marmbrus @rxin @davies This is an example UDAF in JAVA.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/expressions/aggregate2/udaf.scala:
```diff
@@ -0,0 +1,149 @@
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.Row
+
+abstract class UserDefinedAggregateFunction extends Serializable {
```
@marmbrus @rxin @davies This is the UDAF interface.

yhuai commented:
Just added an UDAF interface. Will add comments and doc tomorrow.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37609 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37609/consoleFull) for   PR 7458 at commit [`a19fea6`](https://github.com/apache/spark/commit/a19fea6448879a3fd1e8dc75bc886a78fc0fd2f2).

SparkQA commented:
[Test build #37609 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37609/console) for   PR 7458 at commit [`a19fea6`](https://github.com/apache/spark/commit/a19fea6448879a3fd1e8dc75bc886a78fc0fd2f2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

chenghao-intel commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala:
```diff
@@ -0,0 +1,365 @@
+  extends UnaryNode {
+
+  /** Indicates if this operator is for partial aggregations. */
+  val partialAggregation: Boolean = {
```
Is there any consideration that we don't use the `partialAggregation` passed from `SparkStrategies`?

yhuai replied:
The planner does not need to pass in `partialAggregation`. The operator will figure it out based on modes of those functions.

chenghao-intel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala:
```diff
@@ -0,0 +1,279 @@
+  /**
+   * The offset of this function's buffer in the underlying buffer shared with other functions.
+   */
+  var bufferOffset: Int = 0
```
Instead of explicit expose the `bufferOffset`, probably people prefer using the `BoundReference`, it's created by `BindReferences.bindReference`, I think we can pass in the associated `BoundReference` (bind with bufferSchema) while `initialize` the `AggregateFunction2` in each executor.

We can add the implicit function in this class, to convert `BoundReference` => Int. So people can write code like:
`mutableRow.isNullAt(BoundReference)`

yhuai replied:
Thanks for the suggestion.

`bufferOffset` is an internal thing. For our built-in aggregate functions, we will try to implement them based on `AlgebraicAggregate`, which supports code-generation. The developers of `AlgebraicAggregate` do not need to use `bufferOffset`. For UDAFs, they should be implemented based on `UserDefinedAggregateFunction`. We will hide this offset from UDAFs. So, hopefully developers will not need to use `bufferOffset` at all. 

I will think about this part and see if there is any other way. I will also go through those new interfaces to make sure we properly set their access modifiers.

yhuai committed:
Add comments and doc. Move some classes to the right places.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37653 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37653/consoleFull) for   PR 7458 at commit [`0a827b3`](https://github.com/apache/spark/commit/0a827b3d5162e4e197f37745ef7c2d7a893a3bc4).

SparkQA commented:
[Test build #37653 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37653/console) for   PR 7458 at commit [`0a827b3`](https://github.com/apache/spark/commit/0a827b3d5162e4e197f37745ef7c2d7a893a3bc4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class MyDoubleSum(child: Expression) extends AggregateFunction2 `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Aggregate2Sort(`
  * `case class ConvertAggregateFunction(context: SQLContext) extends Rule[LogicalPlan] `
  * `case class CheckAggregateFunction(context: SQLContext) extends (LogicalPlan => Unit) `
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala:
```diff
@@ -0,0 +1,283 @@
+ */
+private[sql] case object Complete extends AggregateMode
+
+private[sql] case object NoOp extends Expression {
```
We should move this out and comment it.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

Conflicts:
	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala
	sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala

yhuai committed:
Change existing AggregateExpression to AggregateExpression1 and add an AggregateExpression as the common interface for both AggregateExpression1 and AggregateExpression2.

yhuai committed:
Change MyJavaUDAF to MyDoubleSum.

yhuai committed:
Add aggregation iterators.

yhuai committed:
Remove AggregateExpressionSuite.scala since our built-in aggregate functions will be based on AlgebraicAggregate and we need to have another way to test it.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala:
```diff
@@ -0,0 +1,237 @@
+ */
+private[sql] case object Complete extends AggregateMode
+
+private[sql] case object NoOp extends Expression with Unevaluable {
```
Will add comments to it and figure out the right place for it later.

yhuai committed:
Format.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37804 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37804/consoleFull) for   PR 7458 at commit [`3013579`](https://github.com/apache/spark/commit/301357979f9111e3c7106d86c111c275429178e6).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37804 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37804/console) for   PR 7458 at commit [`3013579`](https://github.com/apache/spark/commit/301357979f9111e3c7106d86c111c275429178e6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

yhuai committed:
Support single distinct column set. WIP

yhuai committed:
Handle single distinct column.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

yhuai committed:
newline.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37879 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37879/consoleFull) for   PR 7458 at commit [`85c9c4b`](https://github.com/apache/spark/commit/85c9c4ba4fba7407cdc4e354cfa0a73c88b636ba).

SparkQA commented:
[Test build #37879 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37879/console) for   PR 7458 at commit [`85c9c4b`](https://github.com/apache/spark/commit/85c9c4ba4fba7407cdc4e354cfa0a73c88b636ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedFunction(`
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `case class ConvertAggregateFunction(context: SQLContext) extends Rule[LogicalPlan] `
  * `case class CheckAggregateFunction(context: SQLContext) extends (LogicalPlan => Unit) `
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
Many updates:

1. Implementing COUNT, FIRST, LAST, MAX, MIN, and SUM based on the new interface.
2. Automatically fall back to old aggregation code path if we cannot evaluate the query using the new code path.
3. Refactoring.

yhuai committed:
Enable spark.sql.useAggregate2 by default for testing purpose.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37971 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37971/consoleFull) for   PR 7458 at commit [`8a8ac4a`](https://github.com/apache/spark/commit/8a8ac4a9b5c8c6542e8578d586a0e34c223fdccb).

SparkQA commented:
[Test build #37971 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37971/console) for   PR 7458 at commit [`8a8ac4a`](https://github.com/apache/spark/commit/8a8ac4a9b5c8c6542e8578d586a0e34c223fdccb).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `    case class Data(topic: Vector, index: Int)`
  * `    case class Data(globalTopicTotals: Vector)`
  * `    case class VertexData(id: Long, topicWeights: Vector)`
  * `    case class EdgeData(srcId: Long, dstId: Long, tokenCounts: Double)`
  * `case class UnresolvedFunction(`
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Count(child: Expression) extends AlgebraicAggregate `
  * `case class First(child: Expression) extends AlgebraicAggregate `
  * `case class Last(child: Expression) extends AlgebraicAggregate `
  * `case class Max(child: Expression) extends AlgebraicAggregate `
  * `case class Min(child: Expression) extends AlgebraicAggregate `
  * `case class Sum(child: Expression) extends AlgebraicAggregate `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
Gracefully fallback to old aggregation code path.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37984 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37984/consoleFull) for   PR 7458 at commit [`a3ca551`](https://github.com/apache/spark/commit/a3ca551f77bb1fd0e8de4d4a81ed1f7988da4960).

SparkQA commented:
[Test build #37984 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37984/console) for   PR 7458 at commit [`a3ca551`](https://github.com/apache/spark/commit/a3ca551f77bb1fd0e8de4d4a81ed1f7988da4960).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
Make it compile.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37986 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37986/consoleFull) for   PR 7458 at commit [`00eb298`](https://github.com/apache/spark/commit/00eb2982868bf13b4758c69b11c5a39cfc2d63c2).

SparkQA commented:
[Test build #37986 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37986/console) for   PR 7458 at commit [`00eb298`](https://github.com/apache/spark/commit/00eb2982868bf13b4758c69b11c5a39cfc2d63c2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
bug fix.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38005 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38005/consoleFull) for   PR 7458 at commit [`3b43b24`](https://github.com/apache/spark/commit/3b43b246f26fb4bd6007a4a51ef47291580db37f).

SparkQA commented:
[Test build #38005 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38005/console) for   PR 7458 at commit [`3b43b24`](https://github.com/apache/spark/commit/3b43b246f26fb4bd6007a4a51ef47291580db37f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedFunction(`
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Count(child: Expression) extends AlgebraicAggregate `
  * `case class First(child: Expression) extends AlgebraicAggregate `
  * `case class Last(child: Expression) extends AlgebraicAggregate `
  * `case class Max(child: Expression) extends AlgebraicAggregate `
  * `case class Min(child: Expression) extends AlgebraicAggregate `
  * `case class Sum(child: Expression) extends AlgebraicAggregate `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
Use semanticEquals to replace grouping expressions in the output of the aggregate operator.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38023 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38023/consoleFull) for   PR 7458 at commit [`35b0520`](https://github.com/apache/spark/commit/35b05207dc329173b6c59778830fbbf59752128a).

yhuai commented:
Seems https://github.com/apache/spark/commit/b285ac5ba85fe0b32b00726ad7d3a2efb602e885 did not add golden files for `udf_unhex`. Those files are included in this PR.

yhuai commented:
I will merge this one once it passes jenkins to unblock other work. If you have any comments to this, feel free to leave them at here. I will address them in a follow-up PR.

yhuai committed:
Merge remote-tracking branch 'upstream/master' into UDAF

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38029 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38029/consoleFull) for   PR 7458 at commit [`f1d5901`](https://github.com/apache/spark/commit/f1d5901d44fe47586449b9e7b2ea4cac9be5aa40).

SparkQA commented:
[Test build #38023 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38023/console) for   PR 7458 at commit [`35b0520`](https://github.com/apache/spark/commit/35b05207dc329173b6c59778830fbbf59752128a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedFunction(`
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Count(child: Expression) extends AlgebraicAggregate `
  * `case class First(child: Expression) extends AlgebraicAggregate `
  * `case class Last(child: Expression) extends AlgebraicAggregate `
  * `case class Max(child: Expression) extends AlgebraicAggregate `
  * `case class Min(child: Expression) extends AlgebraicAggregate `
  * `case class Sum(child: Expression) extends AlgebraicAggregate `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

yhuai committed:
Remove unnecessary change.

yhuai committed:
Put the catalyst expression in the comment of the generated code for it.

SparkQA commented:
[Test build #1152 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1152/consoleFull) for   PR 7458 at commit [`7865f5e`](https://github.com/apache/spark/commit/7865f5ef072d1877b73276f28e84afa98c247a82).

AmplabJenkins commented:
Merged build triggered.

yhuai commented:
test this please

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38030 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38030/consoleFull) for   PR 7458 at commit [`7865f5e`](https://github.com/apache/spark/commit/7865f5ef072d1877b73276f28e84afa98c247a82).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #38029 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38029/console) for   PR 7458 at commit [`f1d5901`](https://github.com/apache/spark/commit/f1d5901d44fe47586449b9e7b2ea4cac9be5aa40).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedFunction(`
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Count(child: Expression) extends AlgebraicAggregate `
  * `case class First(child: Expression) extends AlgebraicAggregate `
  * `case class Last(child: Expression) extends AlgebraicAggregate `
  * `case class Max(child: Expression) extends AlgebraicAggregate `
  * `case class Min(child: Expression) extends AlgebraicAggregate `
  * `case class Sum(child: Expression) extends AlgebraicAggregate `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented:
I'm going to merge this epic change!

SparkQA commented:
[Test build #1152 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1152/console) for   PR 7458 at commit [`7865f5e`](https://github.com/apache/spark/commit/7865f5ef072d1877b73276f28e84afa98c247a82).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedFunction(`
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Count(child: Expression) extends AlgebraicAggregate `
  * `case class First(child: Expression) extends AlgebraicAggregate `
  * `case class Last(child: Expression) extends AlgebraicAggregate `
  * `case class Max(child: Expression) extends AlgebraicAggregate `
  * `case class Min(child: Expression) extends AlgebraicAggregate `
  * `case class Sum(child: Expression) extends AlgebraicAggregate `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class FormatString(children: Expression*) extends Expression with ImplicitCastInputTypes `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

SparkQA commented:
[Test build #38030 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38030/console) for   PR 7458 at commit [`7865f5e`](https://github.com/apache/spark/commit/7865f5ef072d1877b73276f28e84afa98c247a82).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedFunction(`
  * `case class Average(child: Expression) extends AlgebraicAggregate `
  * `case class Count(child: Expression) extends AlgebraicAggregate `
  * `case class First(child: Expression) extends AlgebraicAggregate `
  * `case class Last(child: Expression) extends AlgebraicAggregate `
  * `case class Max(child: Expression) extends AlgebraicAggregate `
  * `case class Min(child: Expression) extends AlgebraicAggregate `
  * `case class Sum(child: Expression) extends AlgebraicAggregate `
  * `abstract class AlgebraicAggregate extends AggregateFunction2 with Serializable `
  * `  implicit class RichAttribute(a: AttributeReference) `
  * `trait AggregateExpression1 extends AggregateExpression `
  * `trait PartialAggregate1 extends AggregateExpression1 `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MinFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class MaxFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class CountFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate1 `
  * `case class CollectHashSet(expressions: Seq[Expression]) extends AggregateExpression1 `
  * `case class CombineSetsAndCount(inputSet: Expression) extends AggregateExpression1 `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class AverageFunction(expr: Expression, base: AggregateExpression1)`
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class CombineSum(child: Expression) extends AggregateExpression1 `
  * `case class CombineSumFunction(expr: Expression, base: AggregateExpression1)`
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class SumDistinctFunction(expr: Expression, base: AggregateExpression1)`
  * `case class CombineSetsAndSum(inputSet: Expression, base: Expression) extends AggregateExpression1 `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class FirstFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class LastFunction(expr: Expression, base: AggregateExpression1) extends AggregateFunction1 `
  * `case class Aggregate2Sort(`
  * `case class FinalAndCompleteAggregate2Sort(`
  * `class GroupingIterator(`
  * `class PartialSortAggregationIterator(`
  * `class PartialMergeSortAggregationIterator(`
  * `class FinalSortAggregationIterator(`
  * `class FinalAndCompleteSortAggregationIterator(`
  * `abstract class UserDefinedAggregateFunction extends Serializable `
  * `case class ScalaUDAF(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -0,0 +1,292 @@
+  override def nullable: Boolean = true
+
+  // First is not a deterministic function.
+  override def deterministic: Boolean = false
```
Why did you mark `First` nondeterministic? `First` will give same result for same input all the time, right?

yhuai replied:
First and Last are both non-deterministic. Let's say you have a query like `SELECT key, first(value) FROM table GROUP BY key`. Since value is not ordered, we do not know what result that `first(value)` will return. I think the only case first and last can generate deterministic results are when you use it as a window function and the argument of it is the order by expression, e.g. `SELECT fist(value) OVER (ORDER BY value) FROM table`.

cloud-fan replied:
OK I'm afraid we may need to add a new concept of mutable expression.

The origin motivation that we add nondeterministic concept is we need some special handling for them:

1. stop doing project collapse for nondeterministic expression.
2. stop pushing down filter through project for nondeterministic expression.
3. pull out nondeterministic expression from sort and materialize them.
4. initialize nondeterministic expression before evaluate them.

AFAIK, we need to do these stuff for `rand`, `randn`, `monotonicallyIncreasingId`, which have mutable states inside. Logically `First` is nondeterministic, but we don't need handle it specially, right? Currently we replace old aggregation expression with new ones during physical plan, so all these special handling won't apply to new aggregation expressions, which is not safe in the future.

cc @rxin @marmbrus

marmbrus replied:
Yeah, I think we are abusing the concept of `deterministic` here (though @yhuai you are correct that given the same data it will not always return the same result).  Regardless, we should probably leave that flag for 1-1 functions that are non-deterministic since Aggregates state management is already special cased.

yhuai replied:
@cloud-fan An `Aggregate` operator handles the project list by itself. There is no Project on top of an `Aggregate` to generate results. So, the first two rules you mentioned will not fire.

My question is why we do not need to handle non-deterministic aggregation function specially and why should not use `deterministic` for an aggregate function? Let's say we have a query `sqlContext.sql("select i, f, f, f from (select i, first(j) as f, rand() from t group by i, rand()) tmp")` and let's also assume project collapse works with `Aggregate`. It still makes sense to evaluate first once instead of three times, right? It is probably true that `first` will give you the same answer even if you call it three times, but what about other non-deterministic aggregate functions? Regarding predicate pushdown, we cannot pushdown predicates using the result of an aggregate function through an aggregate operator, right?

yhuai replied:
@cloud-fan We can add comments to explain that `deterministic` of an aggregate function is different from that of a regular 1-1 function like rand. Or, we can leave deterministic as true for aggregate functions. I prefer the first option.
